Control/Monad/Logic/Class.hs

```diff
@@ -45,19 +45,56 @@ class (MonadPlus m) => MonadLogic m where
     --   > do x <- odds `mplus` return 2
     --   >    if even x then return x else mzero
     --
-    --   Such a computation may never consider the 'return 2', and will
-    --   therefore never terminate. By contrast, interleave ensures fair
-    --   consideration of both branches of a disjunction
+    --   Such a computation may never consider the @return 2@, and
+    --   will therefore never return any results. By contrast, using
+    --   'interleave' in place of 'mplus' ensures fair consideration
+    --   of both branches of a disjunction.
+    --
+    --   Note that this sample computation will never terminate: only
+    --   the first value can be safely observed, after which each odd
+    --   value becomes 'mzero' (equivalent to a Prologue @fail@) which
+    --   does not stop the evaluation but indicates there is no value
+    --   to return yet.
+    --
     interleave :: m a -> m a -> m a
 
     -- | Fair conjunction. Similarly to the previous function, consider
-    --   the distributivity law for MonadPlus:
+    --   the distributivity law for 'MonadPlus':
     --
     --   > (mplus a b) >>= k = (a >>= k) `mplus` (b >>= k)
     --
-    --   If 'a >>= k' can backtrack arbitrarily many times, (b >>= k) may never
-    --   be considered. (>>-) takes similar care to consider both branches of
-    --   a disjunctive computation.
+    --   If @(a >>= k)@ can backtrack arbitrarily many times, @(b >>=
+    --   k)@ may never be considered. The '>>-' operator takes similar
+    --   care to consider both branches of a disjunctive computation.
+    --
+    --   Consider the operation:
+    --
+    --   > odds = return 1 `mplus` liftM (2+) odds
+    --   >
+    --   > oddsPlus n = odds >>= \a -> return (a + n)
+    --   >
+    --   > do x <- (return 0 `mplus` return 1) >>= oddsPlus
+    --   >    if even x then return x else mzero
+    --
+    --   This will never produce any values because all values come
+    --   from the @return 1@ driven operation, but the @return 0@ driven
```

Comment:
It does come from the `return 1`  because it starts with odd values so only the `oddsPlus 1` call generates values.

Reply:
It is ambiguous, because this way "all values" refers to results of the whole computation, and "infinite number of potential results" refers to values of `x` in the first line. Initially I read it it vice versa.

Reply:
Ah, you're right.  Let me know if the clarification I pushed helps or if I should work on it more.